File: apple/LC200NumberOfIslands2.java
class Solution {
    public int numIslands(char[][] grid) {
        if (grid.length == 0 || grid[0].length == 0 || grid == null) return 0;
        int answer = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    answer++;
                    bfs(grid, i, j);
                }
            }
        }
        return answer;
    }

    public void bfs(char[][] grid, int i, int j) {
        if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || grid[i][j] == '0') return;
        grid[i][j] = '0';
        bfs(grid, i+1, j);
        bfs(grid, i-1, j);
        bfs(grid, i, j+1);
        bfs(grid, i, j-1);
    }
}

File: Linkedin/LC200NumberOfIslands2.py
class Solution(object):
    # for BFS method, directly will cause TLE, so using extend will be faster, be careful about extend function
    def numIslands(self, grid):
        if not grid or not grid[0]:
            return 0
        m = len(grid)
        n = len(grid[0])
        queue = []
        answer = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    queue.append((i, j))
                    answer += 1
                    while queue:
                        row, col = queue.pop(0)
                        if 0<=row<m and 0<=col<n and grid[row][col] == '1':
                            grid[row][col] = '2'
                            queue.extend([(row+1, col), (row-1, col), (row, col+1), (row, col-1)])
        return answer
